k8s/src/main/scala/io/buoyant/k8s/Watchable.scala

```diff
@@ -138,10 +160,60 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
 
     (_watch(resourceVersion), Closable.ref(close))
   }
+
+  /**
+   * Convert this Watchable into an [[Activity]]
+   * @param resourceVersion whether or not to send the initial resource
+   *                        version of the watched resource. this is a
+   *                        special case due to errors with the
+   *                        ConfigMap interpreter
+   * @param onEvent function called on each [[Watch]] event
+   * TODO: i wish this returned a [[Activity.State]] so that the caller
+   *       of this function could turn a [[Watch.Error]] into an
+   *       [[Activity.Failed]] if they wanted watch errors to fail this
+   *       [[Activity]], although none of our current code exhibits this
+   *       use-case...
+   *         - eliza, 7/18/2017
+   * @return
+   */
+  def activity[T](convert: G => T, resourceVersion: Boolean = true)
+                  (onEvent: (T, W) => T)
+  : Activity[T] =
+    Activity(Var.async[Activity.State[T]](Activity.Pending) { state =>
+      val closeRef = new AtomicReference[Closable](Closable.nop)
+      val pending = get(retryIndefinitely = true)
+        // if the initial GET failed, then the activity is a failure
+        .onFailure { e => state.update(Activity.Failed(e)) }
+        // otherwise, update the activity with the initial state, and
+        // apply the onEvent function to each successive watch event in
+        // the stream
+        .onSuccess { initial =>
+          val initialState = convert(initial)
+          state.update(Activity.Ok(initialState))
+
+          val version = if (resourceVersion) {
+            initial.metadata.flatMap(_.resourceVersion)
+          } else None
+
+          val (stream, close) = watch(None, None, version)
+
+          closeRef.set(close)
+          val _ = stream.scanLeft(initialState) { onEvent }
+            .foreach { s => state.update(Activity.Ok(s)) }
```

Comment:
This is super subtle, but I think the activity needs to updated as part of the first iteration over the stream, rather than in a separate iteration. As written, I don't think the activity will ever update, since the stream doesn't terminate.

I tested this branch on Kubernetes and observed that the dtab was never picking up changes to the config map. Switching back to foldLeft fixes the issue, such that changes in the config map are reflected in the dtab as well.

```
          val _ = stream.foldLeft(initialState) { (s, watchEvent) =>
            val newState = onEvent(s, watchEvent)
            state.update(Activity.Ok(newState))
            newState
          }
```